Combobox: Pressing Escape in while focusing a Combobox inside a Modal will close the Modal

it-vegard opened the pull request:
### Description

If pressing Escape in a Combobox inside of a Modal, the Modal would close.
Combobox should instead swallow the keydown event, so it does not close the Modal.
NB! This fix only swallows the event if FilteredOptions is open.

Fixes #3052 

### Change summary

Suppress keydown event when pressing Escape key while Combobox.FilteredOptions is open

### For discussion:

I don't have any tests yet, because I couldn't get them to fail pre-fix.

Turns out Storybook/Testing Library `userEvent.keyboard("{Escape}")` does not cause the Modal to close, while doing the same manually (even in Storybook) will cause the error. Found [references to the same here](https://github.com/testing-library/user-event/issues/969), and [a potential fix here](https://codesandbox.io/p/sandbox/v14-userevent-keyboard-escape-forked-b9sxol?file=%2Fsrc%2FApp.test.js%3A6%2C1-16%2C3), but then the fix wouldn't work in the test (but it worked in real-life).

Do we need tests for this?

it-vegard committed:
If pressing Escape while FilteredOptions is open, Combobox should swallow the keydown event, so it does not close a Modal

it-vegard committed:
Added changeset

HalvorHaugan commented on @navikt/core/react/src/form/combobox/Input/Input.tsx:
```diff
@@ -132,6 +132,10 @@ const Input = forwardRef<HTMLInputElement, InputProps>(
           }
+        } else if (e.key === "Escape") {
+          if (isListOpen) {
+            e.preventDefault(); // Prevents the Escape on keydown from closing en encasing Modal, as Combobox reacts on keyup.
```
```suggestion
            e.preventDefault(); // Prevents closing an encasing Modal, as Combobox reacts on keyup.
```
Typo + simplification since we already can see from the code that it has to do with Escape keydown.

HalvorHaugan commented on @navikt/core/react/src/form/combobox/Input/Input.tsx:
```diff
@@ -132,6 +132,10 @@ const Input = forwardRef<HTMLInputElement, InputProps>(
             e.preventDefault();
           }
+        } else if (e.key === "Escape") {
+          if (isListOpen) {
```
Should we add `|| value` since Escape also clears the input?

HalvorHaugan commented on @navikt/core/react/src/form/combobox/Input/Input.tsx:
Can we move the code that runs on Escape keyUp to keyDown instead, or is it important that this happens in keyUp? It would be nice to have all the Escape logic in one place.

If not possible, I suggest adding a little comment in `handleKeyUp` right after `case "Escape":`, e.g. "See also handleKeyDown". This should help us remember to update both places when making changes.

it-vegard replied:
I agree that it would make it safer to make changes later to have the code together.

Normally, I am sceptical about doing stuff on keyDown, but it seems (native) Select will also close the dropdown on keyDown, so perhaps it is fine doing that here too. I'll update the code.

it-vegard committed:
Update @navikt/core/react/src/form/combobox/Input/Input.tsx

Co-authored-by: Halvor Haugan <83693529+HalvorHaugan@users.noreply.github.com>

it-vegard committed:
Prevent default Escape from bubbling out of Combobox also when we have a value (which will be cleared).
Also co-located all the Escape-handling in keyDown. This also better mimics Select's native behaviour.

HalvorHaugan commented on @navikt/core/react/src/form/combobox/Input/Input.tsx:
```diff
@@ -132,6 +128,12 @@ const Input = forwardRef<HTMLInputElement, InputProps>(
+        } else if (e.key === "Escape") {
+          if (isListOpen || value) {
+            e.preventDefault(); // Prevents closing an encasing Modal, as Combobox reacts on keyup.
+            clearInput(e);
```
I think `clearInput` should be added to the dependency array.

it-vegard committed:
Add clearInput to dependency array

it-vegard committed:
Legg til et Storybook-eksempel på Combobox i Modal, for manuell testing. La inn en tekst for å forklare caset.

it-vegard committed:
Merge branch 'main' into bugfix/stop-closing-modal-too-when-closing-combobox-with-escape

# Conflicts:
#	@navikt/core/react/src/form/combobox/Input/Input.tsx
#	@navikt/core/react/src/form/combobox/combobox.stories.tsx
